include/aws/s3/s3_client.h

```diff
@@ -487,6 +550,14 @@ struct aws_s3_meta_request_options {
      */
     aws_s3_meta_request_telemetry_fn *telemetry_callback;
 
+    /**
+     * Optional.
+     * Callback for reviewing an upload before it completes.
+     * WARNING: experimental/unstable
+     * See `aws_s3_upload_review_fn`
+     */
+    aws_s3_meta_request_upload_review_fn *upload_review_callback;
```

Comment:
If we going to support single part for the same functionality, how can we cancel the put?

For a single part, there will only be request, and if we invoke the callback as we read through the body, which has already been sent to the server and request has basically finished. We will need to delete the object. Will that be an overkill? 

Just bring it up as the naming is `upload_review`. If we only gonna support MPU for this, we should name it as mpu

Reply:
my think was: to make it work with singlepart, we'd need to very carefully fire this callback as part of the input_stream_read()

Reply:
Or just do checksum calculation when we initially stream data from the user, which would be way simpler, but we'd need benchmarks to prove that it doesn't degrade performance